File: solutions/073-set-matrix-zeroes/solution.py
# 给定一个 m x n 的矩阵，如果一个元素为 0 ，则将其所在行和列的所有元素都设为 0 。请使用原地算法。
# 先遍历一下数组，查看0的位置，对应的行和列标记一下，最后将对应行和列置零
from typing import List


class Solution:
    def set_zeroes(self, matrix: List[List[int]]) -> None:
        m, n = len(matrix), len(matrix[0])
        row = [False] * m
        col = [False] * n
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    row[i] = True
                    col[j] = True
        for i in range(m):
            for j in range(n):
                if row[i] or col[j]:
                    matrix[i][j] = 0


if __name__ == "__main__":
    sol = Solution()
    m = [[1, 1, 1], [1, 0, 1], [1, 1, 1]]
    sol.set_zeroes(m)
    expected = [[1, 0, 1], [0, 0, 0], [1, 0, 1]]
    status = "PASS" if m == expected else "FAIL"
    print(f"{status} | matrix={m}")

File: solutions/073-set-matrix-zeroes/solution.go
package main

import "fmt"

func setZeroes(matrix [][]int) {
	m, n := len(matrix), len(matrix[0])

	firstRowZero := false
	for j := 0; j < n; j++ {
		if matrix[0][j] == 0 {
			firstRowZero = true
			break
		}
	}
	firstColZero := false
	for i := 0; i < m; i++ {
		if matrix[i][0] == 0 {
			firstColZero = true
			break
		}
	}

	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if matrix[i][j] == 0 {
				matrix[i][0] = 0
				matrix[0][j] = 0
			}
		}
	}

	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if matrix[i][0] == 0 || matrix[0][j] == 0 {
				matrix[i][j] = 0
			}
		}
	}

	if firstRowZero {
		for j := 0; j < n; j++ {
			matrix[0][j] = 0
		}
	}
	if firstColZero {
		for i := 0; i < m; i++ {
			matrix[i][0] = 0
		}
	}
}

func equal(a, b [][]int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if len(a[i]) != len(b[i]) {
			return false
		}
		for j := range a[i] {
			if a[i][j] != b[i][j] {
				return false
			}
		}
	}
	return true
}

func main() {
	tests := []struct {
		matrix   [][]int
		expected [][]int
	}{
		{
			[][]int{{1, 0}},
			[][]int{{0, 0}},
		},
		{
			[][]int{{0, 1, 2, 0}, {3, 4, 5, 2}, {1, 3, 1, 5}},
			[][]int{{0, 0, 0, 0}, {0, 4, 5, 0}, {0, 3, 1, 0}},
		},
	}

	for _, tc := range tests {
		setZeroes(tc.matrix)
		status := "PASS"
		if !equal(tc.matrix, tc.expected) {
			status = "FAIL"
		}
		fmt.Printf("%s | matrix=%v\n", status, tc.matrix)
	}
}
